language/stdlib/modules/Roles.move

```diff
@@ -354,5 +354,101 @@ module Roles {
         // Set that the privilege is now put back
         borrow_global_mut<Privilege<Priv>>(owner_address).is_extracted = false;
     }
+
+//**************** Specifications ****************
+
+    /// **Note:** Just started, only a few specs.    
+
+    /// ## Role persistence
+
+    spec module {
+        pragma verify = true;
+    }
+
+    /// Helper functions
+    spec module {
+        define spec_has_role_id(addr: address): bool {
+            exists<RoleId>(addr)
+        }
+
+        define spec_get_role_id(addr: address): u64 {
+            global<RoleId>(addr).role_id
+        }
+
+        define SPEC_ASSOCIATION_ROOT_ROLE_ID(): u64 { 0 }
+        define SPEC_TREASURY_COMPLIANCE_ROLE_ID(): u64 { 1 }
+        define SPEC_DESIGNATED_DEALER_ROLE_ID(): u64 { 2 }
+        define SPEC_VALIDATOR_ROLE_ID(): u64 { 3 }
+        define SPEC_VALIDATOR_OPERATOR_ROLE_ID(): u64 { 4 }
+        define SPEC_PARENT_VASP_ROLE_ID(): u64 { 5 }
+        define SPEC_CHILD_VASP_ROLE_ID(): u64 { 6 }
+        define SPEC_UNHOSTED_ROLE_ID(): u64 { 7 }
+
+    }
+
+    /// Property stated informally at definition of RoleId resource
+    /// TODO: Need a consistent convention for where to document properties.
+    /// For now, restating here.
+    ///
+    /// **Informally:** Once an account at address `A` is granted a role `R` it
+    /// will remain an account with role `R` for all time.    
+    spec schema RoleIdPersists {
+        ensures forall addr: address where old(spec_has_role_id(addr)) :
+            spec_has_role_id(addr) && (old(spec_get_role_id(addr)) == spec_get_role_id(addr));
+    }
+
+    spec module {
+        apply RoleIdPersists to *<T>, *;
+    }
+
+    /// ## Role-specific privileges
+    ///
+    /// Each address has a role-specific privilege corresponding iff that address has
+    /// an RoleId and the role_id of the account matches the privilege.
+    ///
+    spec schema AssociationRootRoleMatchesRoleId {
+        // precondition to grant_root_association
+        // every address has assoc_root role_id iff if it has the privilege
+        // adding one more should preserve it.
+        // TODO BUG (dd): The prover finds many false errors for the following because
+        // it thinks many add_privilege_* functions can put AssociationRootRole on
+        // addresses that don't have the association root RootId.  This is due to
```

Comment:
I'm wondering why we don't use the usual inductive approach here, which says exactly what you state above (namely that AssociationRootRole can only be at address that have RootId role). Because it's a module invariant, it is assumed on entry and does not need to be proven, but just constraints our global states. On exit we have to prove it still holds (with the usual exception of functions which change it).

Reply:
Consider some other module calling say `add_privilege_to_account_treasury_compliance_role<Privilege<AssociationRootRole>>(account, assoc_priv)`, where `account` is indeed a treasury compliance account.

This is entirely legal right now. What we need to add here is the restriction that the type parameter `Priv` for function `add_privilege_to_account_treasury_compliance_role` cannot be an indoor thing like `Privilege<AssociationRootRole>`.

Reply:
I think that's the approach I'm trying to use.

The problem is that the proof fails because add_privilege_designated_dealer appears to violate the property because add_privilege_designated_dealer can add an arbitrary privilege to an address that has the designated_dealer role.  It thinks that arbitrary privilege can be the AssociationRootRole.  In reality, the arbitrary privilege CANNOT be AssociationRootRole because no instance of that resource is accessible to a module that can call add_privilege_designated_dealer.  I think.

Reply:
Well that can be done. We do have the `type<T>()` expression, so you can say such things. I guess it would be a precondition (because there is no matching abort in the program?). Something like:

```
spec add_privilege_to_account_treasury_compliance_role {
   requires type<Priv>() != type<AssociationRootRole>();
}
```

Reply:
Right I think adding this pre condition should(?) solve the issue. It seems that the solution I had in mind was to essentially automate the process by identifying which types `Priv` can't be using some analysis and adding these `requires` automatically.

Reply:
Perhaps the core of the problem is that code actually does not abort on calls like this, and henceforth allows real dangerous operations. If Move cannot abort on concrete type instances for generic parameters, we should ask the language team to add this as a native function. It should be easy, there could be a native function `types_equal<T1, T2>`.

@sblackshear @tzakian

Reply:
I tried to explain better in the comments in the code.

Reply:
Missed some discussion.  I can add the precondition that the arg is not of type AssociationRootRole. 

The prover will catch violations of this in calling modules if we verify them.  I've been trying to avoid assuming that we have verified all calling modules.  When we have an open system, we won't necessarily have access to all calling code.

The Move team should not have to put an assert in there to ensure it, since it's true.  The information exists for us to prove, just looking at the Roles module, that add_privilege_to.... cannot have an AssociationRootRole as an argument.

As a work-around, a precondition might be fair game, though.